internal/messagequeue/donthavetimeoutmgr.go

```diff
@@ -259,28 +268,37 @@ func (dhtm *dontHaveTimeoutMgr) checkForTimeouts() {
 	// Schedule the next check for the moment when the oldest pending want will
 	// timeout
 	oldestStart := dhtm.wantQueue[0].sent
-	until := time.Until(oldestStart.Add(dhtm.timeout))
+	until := oldestStart.Add(dhtm.timeout).Sub(dhtm.clock.Now())
 	if dhtm.checkForTimeoutsTimer == nil {
-		dhtm.checkForTimeoutsTimer = time.AfterFunc(until, func() {
-			dhtm.lk.Lock()
-			defer dhtm.lk.Unlock()
-
-			dhtm.checkForTimeouts()
-		})
+		dhtm.checkForTimeoutsTimer = dhtm.clock.Timer(until)
+		go dhtm.consumeTimeouts()
 	} else {
 		dhtm.checkForTimeoutsTimer.Stop()
 		dhtm.checkForTimeoutsTimer.Reset(until)
 	}
 }
 
+func (dhtm *dontHaveTimeoutMgr) consumeTimeouts() {
+	for {
```

Comment:
I don't think this should be a for loop, because checkForTimeouts() is a recursive call

Reply:
note this will only get called once cause then checkForTimeoutsTimer will not be null.

Reply:
I think that the for loop will continue looping forever though - so I'm suggesting that we only have the select (without putting it inside a for loop)